### PR TITLE
Rewrite System.Text.Json stream tests to be async friendly and enable on WASM

### PR DESCRIPTION
**Pull request by akoeplinger**
The tests are using a (De)SerializationWrapper so the same code can be used both for String and Stream types.
It does that by wrapping the async Stream serialization calls in `Task.Run().GetAwaiter().GetResult()` to turn them into sync calls.
However that doesn't work on WebAssembly since we can't wait on tasks as there's only a single thread.

To fix this inverse the wrapper so the synchronous String calls are turned into async and use normal awaits for the Stream calls.

This allows the test suite to pass on WebAssembly: `Tests run: 8349, Errors: 0, Failures: 0, Skipped: 11. Time: 475.528706s`

### CODE REVIEW

**Comment by stephentoub** [comment]
We can skip fewer now, right? :)

**Comment by akoeplinger** [comment]
Yeah I'm working on it :)

File: src/libraries/System.Text.Json/tests/Utf8JsonReaderTests.cs [hunk 1]
```diff
@@ -1626,6 +1627,11 @@ public static void TestPartialJsonReaderSlicesSpecialNumbers(TestCaseType type,
         [InlineData(512)]
         public static void TestDepth(int depth)
         {
+            if (PlatformDetection.IsBrowser && depth >= 256)
```
Comment (marek-safar) [inline src/libraries/System.Text.Json/tests/Utf8JsonReaderTests.cs:1630]: This should check interpreter not browser

File: src/libraries/System.Text.Json/tests/Utf8JsonReaderTests.MultiSegment.cs [hunk 2]
```diff
@@ -486,6 +486,12 @@ public static void TestJsonReaderUtf8SegmentSizeOne(bool compactData, TestCaseTy
         [MemberData(nameof(LargeTestCases))]
         public static void TestJsonReaderLargeUtf8SegmentSizeOne(bool compactData, TestCaseType type, string jsonString)
         {
+            // Skipping really large JSON on Browser to prevent OOM
+            if (PlatformDetection.IsBrowser && (type == TestCaseType.Json40KB || type == TestCaseType.Json400KB || type == TestCaseType.ProjectLockJson))
```
Comment (marek-safar) [inline src/libraries/System.Text.Json/tests/Utf8JsonReaderTests.MultiSegment.cs:490]: nit: use switch

Comment (akoeplinger) [reply src/libraries/System.Text.Json/tests/Utf8JsonReaderTests.MultiSegment.cs:490]: There's existing code that uses the same pattern so I think it's better to be consistent.

**Comment by akoeplinger** [comment]
@stephentoub this is ready for review now :)

**Comment by ericstj** [comment]
@steveharter @layomia PTAL

File: src/libraries/System.Text.Json/tests/JsonDocumentTests.cs [hunk 3]
```diff
@@ -3650,7 +3647,7 @@ private static string GetCompactJson(TestCaseType testCaseType, string jsonStrin
             return s_compactJson[testCaseType] = existing;
         }
 
-        [Fact]
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsThreadingSupported))]
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/tests/JsonDocumentTests.cs:3650]: It wouldn't be testing the exact same thing, but it seems like this test would "just work" if you changed the `Task.WaitAll(tasks)` to instead be `await Task.WhenAll(tasks)` and made the method async, no?

File: src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Cache.cs [hunk 4]
```diff
@@ -21,35 +22,35 @@ public void MultipleThreadsLooping()
             }
         }
 
-        [Fact]
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsThreadingSupported))]
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Cache.cs:25]: Same comment here with the WaitAll.  It seems like it's still a useful test, as even if the code won't run in parallel, it could still run interleaved.

File: src/libraries/System.Text.Json/tests/JsonDocumentTests.cs (hunk 3 above)
Comment (akoeplinger) [reply src/libraries/System.Text.Json/tests/JsonDocumentTests.cs:3650]: Yeah that does indeed work. I can send a followup PR but I'd like to first get this one in while CI is green :)

File: src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.AttributePresence.cs [hunk 5]
```diff
@@ -2,159 +2,160 @@
...
+            await RunTest<PrivateParameterlessConstructor_PublicParameterizedCtor_WithAttribute>();
         }
 
         [Fact]
-        public void Class_MultiplePublicParameterizedCtors_NoPublicParameterlessCtor_NoAttribute_NotSupported()
+        public async Task Class_MultiplePublicParameterizedCtors_NoPublicParameterlessCtor_NoAttribute_NotSupported()
         {
-            Assert.Throws<NotSupportedException>(() => Serializer.Deserialize<MultiplePublicParameterizedCtor>(@"{""MyInt"":1,""MyString"":""1""}"));
+            await Assert.ThrowsAsync<NotSupportedException>(() => Serializer.DeserializeAsync<MultiplePublicParameterizedCtor>(@"{""MyInt"":1,""MyString"":""1""}"));
         }
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.AttributePresence.cs:83]: nit: remove `async` and return `Assert.ThrowAsync` instead.

File: src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Cache.cs [hunk 6]
```diff
@@ -137,9 +138,9 @@ void Serialize<T>(object[] args)
                 string json = JsonSerializer.Serialize(localTestObj, s_options);
             };
 
-            void Deserialize<T>(string json)
+            async Task Deserialize<T>(string json)
             {
-                ITestClass obj = (ITestClass)Serializer.Deserialize<T>(json, s_options);
+                ITestClass obj = (ITestClass)await Serializer.DeserializeAsync<T>(json, s_options);
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Cache.cs:143]: Is this needed if you are excluding the test from single-threading platforms?

File: src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Cache.cs [hunk 7]
```diff
@@ -113,18 +114,18 @@ public void PropertyCacheWithMinInputsLast()
...
-            Serializer.Deserialize<ClassWithConstructor_SimpleAndComplexParameters>(json, options);
+            await Serializer.DeserializeAsync<ClassWithConstructor_SimpleAndComplexParameters>(json, options);
         }
 
         // Use a common options instance to encourage additional metadata collisions across types. Also since
         // this options is not the default options instance the tests will not use previously cached metadata.
         private JsonSerializerOptions s_options = new JsonSerializerOptions();
 
-        [Fact]
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsThreadingSupported))]
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Cache.cs:128]: Same comment as https://github.com/dotnet/runtime/pull/38663/files#r452481795

File: src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Cache.cs [hunk 8]
```diff
@@ -21,35 +22,35 @@ public void MultipleThreadsLooping()
...
         public void MultipleThreads()
         {
             // Verify the test class has >32 properties since that is a threshold for using the fallback dictionary.
             Assert.True(typeof(ClassWithConstructor_SimpleAndComplexParameters).GetProperties(BindingFlags.Instance | BindingFlags.Public).Length > 32);
 
-            void DeserializeObject(string json, Type type, JsonSerializerOptions options)
+            async Task DeserializeObject(string json, Type type, JsonSerializerOptions options)
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Cache.cs:31]: nit, for this and subsequent local functions:
```suggestion
            async Task DeserializeObjectAsync(string json, Type type, JsonSerializerOptions options)
```

File: src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.ParameterMatching.cs [hunk 9]
```diff
@@ -28,191 +29,191 @@ public ConstructorTests(DeserializationWrapper serializer)
...
+            Assert.Null(await Serializer.DeserializeAsync<Point_3D>("null"));
         }
 
         [Fact]
-        public void JsonExceptionWhenAssigningNullToStruct()
+        public async Task JsonExceptionWhenAssigningNullToStruct()
         {
-            Assert.Throws<JsonException>(() => Serializer.Deserialize<Point_2D_With_ExtData>("null"));
+            await Assert.ThrowsAsync<JsonException>(() => Serializer.DeserializeAsync<Point_2D_With_ExtData>("null"));
         }
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.ParameterMatching.cs:42]: Remove `async` and return `Assert.ThrowsAsync` instead or perhaps we should add more assertions? The former should not be part of this PR.

File: src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Stream.cs [hunk 10]
```diff
@@ -28,36 +28,31 @@ async Task RunTest<T>(byte[] testData)
...
-            tasks[10] = Task.Run(async () => await RunTest<Point_MembersHave_JsonIgnore>(Point_MembersHave_JsonIgnore.s_data));
+            await RunTest<Point_MembersHave_JsonPropertyName>(Point_MembersHave_JsonPropertyName.s_data);
+            await RunTest<Point_MembersHave_JsonConverter>(Point_MembersHave_JsonConverter.s_data);
+            await RunTest<Point_MembersHave_JsonIgnore>(Point_MembersHave_JsonIgnore.s_data);
             // Complex JSON as last argument works
-            tasks[11] = Task.Run(async () => await RunTest<Point_With_Array>(Point_With_Array.s_data));
-            tasks[12] = Task.Run(async () => await RunTest<Point_With_Dictionary>(Point_With_Dictionary.s_data));
-            tasks[13] = Task.Run(async () => await RunTest<Point_With_Object>(Point_With_Object.s_data));
-
-            Task.WaitAll(tasks);
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Stream.cs:56]: Can you also apply the pattern suggested in https://github.com/dotnet/runtime/pull/38663/files#r452481795 here?

File: src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Stream.cs [hunk 11]
```diff
@@ -75,36 +70,31 @@ async Task RunTest<T>(string testData)
...
-            tasks[10] = Task.Run(async () => await RunTest<Point_MembersHave_JsonIgnore>(Point_MembersHave_JsonIgnore.s_json));
+            await RunTest<Point_MembersHave_JsonPropertyName>(Point_MembersHave_JsonPropertyName.s_json);
+            await RunTest<Point_MembersHave_JsonConverter>(Point_MembersHave_JsonConverter.s_json);
+            await RunTest<Point_MembersHave_JsonIgnore>(Point_MembersHave_JsonIgnore.s_json);
             // Complex JSON as last argument works
-            tasks[11] = Task.Run(async () => await RunTest<Point_With_Array>(Point_With_Array.s_json));
-            tasks[12] = Task.Run(async () => await RunTest<Point_With_Dictionary>(Point_With_Dictionary.s_json));
-            tasks[13] = Task.Run(async () => await RunTest<Point_With_Object>(Point_With_Object.s_json));
-
-            Task.WaitAll(tasks);
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Stream.cs:103]: Same here.

File: src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Stream.cs (hunk 10 above)
Comment (Jozkee) [reply src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Stream.cs:56]: Isn't the intent of this test to detect race conditions? By awaiting one-by-one doesn't defy said purpose?

File: src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Stream.cs [hunk 12]
```diff
@@ -140,12 +130,8 @@ async Task RunTest<T>()
                 }
             }
 
-            Task[] tasks = new Task[2];
-
-            tasks[0] = Task.Run(async () => await RunTest<Class_With_Ctor_With_65_Params>());
-            tasks[1] = Task.Run(async () => await RunTest<Struct_With_Ctor_With_65_Params>());
-
-            Task.WaitAll(tasks);
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Stream.cs:148]: Same here.

File: src/libraries/System.Text.Json/tests/Serialization/Stream.ReadTests.cs [hunk 13]
```diff
@@ -20,7 +20,7 @@ public static async Task ReadNullArgumentFail()
             await Assert.ThrowsAsync<ArgumentNullException>(async () => await JsonSerializer.DeserializeAsync(new MemoryStream(), (Type)null));
         }
 
-        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsThreadingSupported))]
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/tests/Serialization/Stream.ReadTests.cs:23]: Why this used to be disabled?

File: src/libraries/System.Text.Json/tests/Utf8JsonReaderTests.cs [hunk 14]
```diff
@@ -1609,7 +1610,7 @@ public static void TestPartialJsonReaderSlicesSpecialNumbers(TestCaseType type,
             }
         }
 
-        [Theory]
+        [ConditionalTheory]
```
Comment (Jozkee) [inline src/libraries/System.Text.Json/tests/Utf8JsonReaderTests.cs:1613]: What's the point of a `ConditionalTheory` without parameters?

File: src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Cache.cs (hunk 7 above)
Comment (stephentoub) [inline src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Cache.cs:128]: Same comment on this one

**Review by stephentoub — APPROVED** [review]
Other than my comments, LGTM.

**Review by akoeplinger — COMMENTED** [review]
@Jozkee @stephentoub I addressed your feedback, please take another look :)

File: src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Stream.cs [hunk 15]
```diff
@@ -129,8 +139,12 @@ async Task RunTest<T>()
...
             }
 
-            await RunTest<Class_With_Ctor_With_65_Params>();
-            await RunTest<Struct_With_Ctor_With_65_Params>();
+            Task[] tasks = new Task[2];
+
+            tasks[0] = Task.Run(async () => await RunTestAsync<Class_With_Ctor_With_65_Params>());
+            tasks[1] = Task.Run(async () => await RunTestAsync<Struct_With_Ctor_With_65_Params>());
+
+            await Task.WhenAll(tasks);
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Stream.cs:147]: You made these run in parallel just for fun? :)

File: src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Stream.cs [hunk 16]
```diff
@@ -69,33 +74,38 @@ async Task RunTest<T>(string testData)
...
+            tasks[10] = Task.Run(async () => await RunTestAsync<Point_MembersHave_JsonIgnore>(Point_MembersHave_JsonIgnore.s_json));
             // Complex JSON as last argument works
-            await RunTest<Point_With_Array>(Point_With_Array.s_json);
-            await RunTest<Point_With_Dictionary>(Point_With_Dictionary.s_json);
-            await RunTest<Point_With_Object>(Point_With_Object.s_json);
+            tasks[11] = Task.Run(async () => await RunTestAsync<Point_With_Array>(Point_With_Array.s_json));
+            tasks[12] = Task.Run(async () => await RunTestAsync<Point_With_Dictionary>(Point_With_Dictionary.s_json));
+            tasks[13] = Task.Run(async () => await RunTestAsync<Point_With_Object>(Point_With_Object.s_json));
+
+            await Task.WhenAll(tasks);
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Stream.cs:102]: Same question; is there a benefit of introducing parallelism here?  Presumably it would make this test harder to debug. and it doesn't appear to be focused on concurrency issues.

File: src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Stream.cs [hunk 17]
```diff
@@ -27,33 +27,38 @@ async Task RunTest<T>(byte[] testData)
...
+            tasks[10] = Task.Run(async () => await RunTestAsync<Point_MembersHave_JsonIgnore>(Point_MembersHave_JsonIgnore.s_data));
             // Complex JSON as last argument works
-            await RunTest<Point_With_Array>(Point_With_Array.s_data);
-            await RunTest<Point_With_Dictionary>(Point_With_Dictionary.s_data);
-            await RunTest<Point_With_Object>(Point_With_Object.s_data);
+            tasks[11] = Task.Run(async () => await RunTestAsync<Point_With_Array>(Point_With_Array.s_data));
+            tasks[12] = Task.Run(async () => await RunTestAsync<Point_With_Dictionary>(Point_With_Dictionary.s_data));
+            tasks[13] = Task.Run(async () => await RunTestAsync<Point_With_Object>(Point_With_Object.s_data));
+
+            await Task.WhenAll(tasks);
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Stream.cs:55]: Same for all of these.

**Review by stephentoub — APPROVED** [review]
Other than not being clear on why you added concurrency into some of the tests, LGTM.

File: src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Stream.cs (hunk 15 above)
Comment (akoeplinger) [reply src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Stream.cs:147]: These were actually already running in parallel before this PR.

File: src/libraries/System.Text.Json/tests/Serialization/CacheTests.cs [hunk 18]
```diff
@@ -121,7 +121,7 @@ void Test(int i)
                 tasks[i + 1] = Task.Run(() => Test(TestClassCount - 2));
             }
 
-            Task.WaitAll(tasks);
+            await Task.WhenAll(tasks);
```
Comment (steveharter) [inline src/libraries/System.Text.Json/tests/Serialization/CacheTests.cs:124]: FYI this area of code really does care about threading, and the changes shouldn't affect the multi-threading aspects.

File: src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.AttributePresence.cs [hunk 19]
```diff
@@ -1,159 +1,160 @@
...
+using System.Threading.Tasks;
 using Xunit;
 
 namespace System.Text.Json.Serialization.Tests
 {
     public abstract partial class ConstructorTests
     {
         [Fact]
-        public void NonPublicCtors_NotSupported()
+        public async Task NonPublicCtors_NotSupported()
```
Comment (steveharter) [inline src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.AttributePresence.cs:12]: Why was this class modified? I don't see any Task calls in the original code.

File: src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Cache.cs [hunk 20]
```diff
@@ -9,46 +9,47 @@ namespace System.Text.Json.Serialization.Tests
...
+        public async Task MultipleThreads()
         {
             // Verify the test class has >32 properties since that is a threshold for using the fallback dictionary.
             Assert.True(typeof(ClassWithConstructor_SimpleAndComplexParameters).GetProperties(BindingFlags.Instance | BindingFlags.Public).Length > 32);
 
-            void DeserializeObject(string json, Type type, JsonSerializerOptions options)
+            async Task DeserializeObjectAsync(string json, Type type, JsonSerializerOptions options)
             {
-                var obj = Serializer.Deserialize(json, type, options);
+                var obj = await Serializer.DeserializeAsync(json, type, options);
```
Comment (steveharter) [inline src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Cache.cs:32]: Why was this changed to use the Async version of the API? The original intent was to test non-async APIs for thread safety.

Comment (stephentoub) [reply src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Cache.cs:32]: The original test was calling a synchronous method which was then overridden by two different implementations, one that used the underlying sync API and one that used the underlying async API and then blocked waiting on it.  This change inverts that, so it's an async API that then uses the underlying sync API and wraps it in a task or the underlying async API.  It's done to avoid sync-over-async, which blows up on wasm.

File: src/libraries/System.Text.Json/tests/Serialization/DeserializationWrapper.cs [hunk 21]
```diff
@@ -16,55 +16,49 @@ public abstract class DeserializationWrapper
...
-            protected internal override T Deserialize<T>(string json, JsonSerializerOptions options = null)
+            protected internal override async Task<T> DeserializeAsync<T>(string json, JsonSerializerOptions options = null)
             {
                 if (options == null)
                 {
                     options = _optionsWithSmallBuffer;
                 }
 
-                return Task.Run(async () =>
+                using (MemoryStream stream = new MemoryStream(Encoding.UTF8.GetBytes(json)))
```
Comment (steveharter) [inline src/libraries/System.Text.Json/tests/Serialization/DeserializationWrapper.cs:45]: So this is the key area of code where we switch the wrapper to use await...

File: src/libraries/System.Text.Json/tests/Serialization/SerializationWrapper.cs [hunk 22]
```diff
@@ -19,96 +19,90 @@ public abstract class SerializationWrapper
...
-            protected internal override string Serialize(object value, Type inputType, JsonSerializerOptions options = null)
+            protected internal override async Task<string> SerializeAsync(object value, Type inputType, JsonSerializerOptions options = null)
             {
-                return Task.Run(async () =>
-                {
-                    using var stream = new MemoryStream();
-                    await JsonSerializer.SerializeAsync(stream, value, inputType, options);
-                    return Encoding.UTF8.GetString(stream.ToArray());
-                }).GetAwaiter().GetResult();
+                using var stream = new MemoryStream();
```
Comment (steveharter) [inline src/libraries/System.Text.Json/tests/Serialization/SerializationWrapper.cs:59]: Again, this is the key area of code where we switch the wrapper to use await without Task.Run.

**Comment by steveharter** [comment]
One global comment is that the renaming of `Serialize` to `SerializeAsync` and `Deserialize` to `DeserializeAsync` is now a bit confusing since it reads as being async when in the majority of cases it isn't. I realize we can't remove the 'await' modifier but perhaps these forms are more understandable:
`await Serializer.Serialize(`
`await Serializer.SerializeSyncOrAsync(`
`await Serializer.SerializeWrapper(`

File: src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Cache.cs (hunk 20 above)
Comment (akoeplinger) [reply src/libraries/System.Text.Json/tests/Serialization/ConstructorTests/ConstructorTests.Cache.cs:32]: Correct. I've tried to capture this in the PR description but please let me know if I could make this clearer.

**Comment by akoeplinger** [comment]
@steveharter I switched to using `Serializer.SerializeWrapper()` instead of the Async suffix

**Review by steveharter — APPROVED** [review]
